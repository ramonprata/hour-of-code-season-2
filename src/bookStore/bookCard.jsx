import React from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import FavoriteIcon from '@material-ui/icons/Favorite';
import { ShoppingCart } from '@material-ui/icons/';
import { Grid, Typography, IconButton, CardMedia } from '@material-ui/core';
import { setShoppingCart, setShoppingWishList } from './bookStoreActions';
import { goToPreview } from '../shared/utils';
export const bookCard = props => {
  const { classes, book, setShoppingCart, setShoppingWishList } = props;
  return (
    <Card className={classes.card}>
      <Grid container direction="column" justify="flex-end">
        <CardMedia
          id={`imgBook${book.id}`}
          onClick={() => goToPreview(book)}
          className={classes.cover}
          image={book.volumeInfo.imageLinks.thumbnail}
        />
        <CardContent className={classes.cardContent}>
          <Typography component="p" variant="subtitle1" id={`bookTitle${book.id}`}>
            {book.volumeInfo.title}
          </Typography>
          <Typography variant="subtitle2" color="textSecondary" id={`bookAuthors${book.id}`}>
            {book.volumeInfo.authors ? book.volumeInfo.authors.join(' ') : ''}
          </Typography>
        </CardContent>
        <div>
          <IconButton
            aria-label="Add to favorites"
            id={`btnFavorite${book.id}`}
            onClick={() => setShoppingWishList(book)}
          >
            <FavoriteIcon />
          </IconButton>
          <IconButton
            aria-label="Share"
            id={`btnShopping${book.id}`}
            onClick={() => setShoppingCart(book)}
          >
            <ShoppingCart />
          </IconButton>
        </div>
      </Grid>
    </Card>
  );
};

bookCard.propTypes = {
  classes: PropTypes.object.isRequired
};

bookCard.defaultProps = {
  book: {
    volumeInfo: { authors: [], title: '' }
  }
};

const styles = theme => ({
  card: {
    height: 350,
    width: 250,
    cursor: 'pointer',
    margin: '20px 0'
  },
  cardContent: {
    height: 75
  },
  cover: {
    height: 0.55 * 350,
    width: 250
  }
});

const mapStateToProps = ({ bookStore }) => {
  return {
    bookTitle: bookStore.title
  };
};

const mapDispatchToProps = {
  setShoppingCart: setShoppingCart,
  setShoppingWishList: setShoppingWishList
};

export default compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  withStyles(styles)
)(bookCard);
