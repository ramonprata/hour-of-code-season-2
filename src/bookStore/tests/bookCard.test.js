import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { bookCard as BookCard } from '../bookCard';
import { booksMock } from './booksMock';
configure({ adapter: new Adapter() });
/**
 * shallow: Função do Enzyme para conseguir montar e manipular nossos componentes em tempo de teste
 * configure: Função do Enzyme para "setar" versão do adapter do Enzyme para cada versão do React
 */

describe('Testa <BookCard />', function() {
  // moca as funções que o bookcard recebe como props
  // assim podemos fazer asserts sobre as funções
  let setShoppingCartMock = jest.fn();
  let setShoppingWishListMock = jest.fn();
  const bookMock = { ...booksMock.data[0] };
  let component;

  // Antes de cada teste, cria um componente BookCard
  beforeEach(() => {
    component = shallow(
      <BookCard
        setShoppingCart={setShoppingCartMock}
        setShoppingWishList={setShoppingWishListMock}
        classes={{}}
        book={bookMock}
      />
    );
  });

  // A função toMatchSnapshot() irá criar um Snapshot do componente
  // em formato JSON. Assim, se o componente for alterado o teste de snapshot
  // não irá passar ou seja, irá evidenciar a alteração do componente para altermos os testes
  it('Deve renderizar <BookCard /> ', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it(`Deve renderizar a imagem do book: imgBook${bookMock.id}`, () => {
    // encontra a imagem pelo id
    const imagemBook = component.find(`#imgBook${bookMock.id}`);
    // valida se a imagem existe
    expect(imagemBook.exists()).toEqual(true);
    // valida se a imagem renderizada corresponde a imagem do book passado como props
    expect(imagemBook.props().image).toEqual(bookMock.volumeInfo.imageLinks.thumbnail);
  });

  it(`Ao clicar na imagem do <BookCard /> deve chamar o window.open()
    E abrir uma aba com o preview do book`, () => {
    // mocka a função window.open() do browser
    global.open = jest.fn();
    // encontra a imagem
    const imagemBook = component.find(`#imgBook${bookMock.id}`);
    // simula o click na imagem
    imagemBook.simulate('click');
    // valida se a função open foi chamada
    expect(global.open).toHaveBeenCalled();
    // valida se a função open recebeu o link do preview do book como parâmetro
    expect(global.open).toBeCalledWith(bookMock.volumeInfo.previewLink);
  });

  it(`Deve renderizar o botão de incluir em na lista e desejos: btnFavorite${bookMock.id}`, () => {
    // valida se o botão de favoritar o book foi renderizado
    expect(component.find(`#btnFavorite${bookMock.id}`).exists()).toEqual(true);
  });

  it(`Ao clicar no de icone de  incluir em na lista e desejos: btnFavorite${
    bookMock.id
  } a função props.setShoppingWishList() deve ser executada `, () => {
    // encontra o botão
    const botao = component.find(`#btnFavorite${bookMock.id}`);
    // simula o click
    botao.simulate('click');
    // verifica se a função setShoppingWishListMock() foi executada
    expect(setShoppingWishListMock).toHaveBeenCalled();
  });

  it(`Deve renderizar o botão de incluir no carrinho de compras: btnShopping${bookMock.id}`, () => {
    // valida se o botão de incluir no carrinho o book foi renderizado
    expect(component.find(`#btnShopping${bookMock.id}`).exists()).toEqual(true);
  });

  it(`Ao clicar no botão  de incluir no carrinho de compras: btnShopping${
    bookMock.id
  } a função props.setShoppingCart() deve ser executada `, () => {
    // encontra o botão
    const botao = component.find(`#btnShopping${bookMock.id}`);
    // simula o click
    botao.simulate('click');
    // verifica se a função setShoppingWishListMock() foi executada
    expect(setShoppingCartMock).toHaveBeenCalled();
  });

  it(`Deve renderizar o titulo do book ${bookMock.volumeInfo.title}`, () => {
    // Encontra o componente que renderiza o titulo do book: Typography Material UI
    expect(component.find(`#bookTitle${bookMock.id}`).exists()).toEqual(true);
    // Valida se o children do Typography renderiza o titulo do book passado como props
    expect(
      component
        .find(`#bookTitle${bookMock.id}`)
        .children()
        .text()
    ).toEqual(bookMock.volumeInfo.title);
  });

  it(`Deve renderizar os autores do book`, () => {
    // monta mock de autores do book para validar
    const autoresDoLivroMock = `${bookMock.volumeInfo.authors.join(' ')}`;
    // Encontra o componente que renderiza os autores do book: Typography Material UI
    expect(component.find(`#bookAuthors${bookMock.id}`).exists()).toEqual(true);
    // Valida se o children do Typography renderiza os autores do book passado como props
    expect(
      component
        .find(`#bookAuthors${bookMock.id}`)
        .children()
        .text()
    ).toEqual(autoresDoLivroMock);
  });
});
