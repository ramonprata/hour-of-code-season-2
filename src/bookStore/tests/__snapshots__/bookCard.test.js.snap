// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testa <BookCard /> Deve renderizar <BookCard />  1`] = `
<WithStyles(ForwardRef(Card))>
  <WithStyles(ForwardRef(Grid))
    container={true}
    direction="column"
    justify="flex-end"
  >
    <WithStyles(ForwardRef(CardMedia))
      id="imgBookH9lBCwAAQBAJ"
      image="http://books.google.com/books/content?id=H9lBCwAAQBAJ&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api"
      onClick={[Function]}
    />
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        component="p"
        id="bookTitleH9lBCwAAQBAJ"
        variant="subtitle1"
      >
        Jane Eyre
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Typography))
        color="textSecondary"
        id="bookAuthorsH9lBCwAAQBAJ"
        variant="subtitle2"
      >
        Charlotte BrontÃ«
      </WithStyles(ForwardRef(Typography))>
    </WithStyles(ForwardRef(CardContent))>
    <div>
      <WithStyles(ForwardRef(IconButton))
        aria-label="Add to favorites"
        id="btnFavoriteH9lBCwAAQBAJ"
        onClick={[Function]}
      >
        <FavoriteIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="Share"
        id="btnShoppingH9lBCwAAQBAJ"
        onClick={[Function]}
      >
        <ShoppingCartIcon />
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Card))>
`;
